Handle Metadata in `io.ascii` MRT Tables

vandalt opened the pull request:
### Description

This pull request aims to address reading and writing metadata in MRT tables with the `io.ascii` sub-package. It is still a draft, but I had some questions regarding the implementation (see "Questions" below)

Currently, only reading is implemented.  It is done via the `update_meta()` method for `MrtHeader`. The method parses all metadata fields (`Title`, `Authors`, `Table`, and all `Notes ({n})` and store their content as list of lines before converting it to its final format (see "What format should be used for metadata?" below).


### Questions

Feedback welcome from everyone, but CC-ing @hamogu and @dhomeier who participated in the discussion on #16787.

**What format should be used for metadata?**

Currently, I'm using

- `Table.meta` to store the metadata
- `Table.meta["top"]` to store the "top" metadata (Title, Authors, Table), as an `OrderedDict`
- `Table.meta["notes"]` to store notes, as a list.

The structure of the metadata values are either:

- `OrderedDict` for notes that map string flags to their descriptions (e.g. note 2 in [cds2.dat](https://github.com/astropy/astropy/blob/b8bc2caa30cc3a0cef19166d0212c2933122c85d/astropy/io/ascii/tests/data/cds2.dat))
- `list`, for Authors
- strings for everything else, where lines are joined by a single space

Let me know if you have comments or suggestions about this.

**How should continuation be handled?**

When a given metadata field is split over multiple lines in the input file, should it be kept this way in the metadata? Currently it is not: lines are concatenated and will be split at the maximum length when writing back. This means that the format is not 1-to-1 when round-tripping, but the content is not altered, only the position of line splits. If we want to preserve the exact position of line splits, we could 1) store metadata as a list of lines or 2) concatenate lines with `\n` instead of a space. In both cases, this would not blend well with "structured" metadata (author list and  "mapping" notes, see 2 and 3 in the description above). Another option would be to keep track of the split locations as "extra" metadata and then apply that when writing the table back. If feel like the complexity might not be worth the small consistency improvement, especially that currently long byte-by-byte descriptions can have different split positions, and the format is not exactly 1-to-1 already (e.g. limits information is added even when not present in the input).


### TODO

Things that should be done before this is ready for review:

- [x] Implement writing metadata
- [ ] Add better tests (currently only a very basic test is implemented)

### Additional context

It is related to #12039, which started to  implement _writing_ MRT metadata. However, that PR was written before MRT and CDS were split in separate modules. I started with reading as I felt this was easier to implement and would help decide which data format to use. There also seems to be some overlap with #14883.

Fixes #16787.

vandalt committed:
Add failing test for metadata in cds.dat [ci skip]

vandalt committed:
Add `update_metadata` to `MrtHeader` (WIP)

vandalt committed:
Add test to check that two lines of MRT title were merged in metadata

vandalt committed:
Handle "top" metadata keys all at once for MRT headers

vandalt committed:
Properly parse multi-line fields in MRT metadata (WIP)

It works, but the tests are not extensive and the code could probably be
more concise

vandalt committed:
Move code that parses metadata lines to separate function

Will be re-usable for "map" notes.
Current function stores metadata as list of lines instead of strings or
structured data.

vandalt committed:
[ci skip] Handle structured metadata (author list and "mapping notes")

First parse all metadata to lists of strings,
then handle mapped notes -> odict,
then concatenate all other metadata to strings,
then split author list into Python list

vandalt committed:
Rename `test_roundtrip_mrt_table` to `test_roundtrip_cds_table`

The test uses a CDS table but not MRT-compatible

vandalt committed:
Write MRT data without wrapping and test that runs the `write` function

vandalt committed:
[ci skip] Wrap MRT metadata when writing (WIP)

"Mapping notes" are wrapped individually and then all indented.
Continuation always uses indentation of 4 spaces
Code neads to be cleaned up. This is first working version.
Still needs proper testing

vandalt commented:
I just finished a first attempt at writing the metadata.

- The current implementation does not keep track of how the lines are split in the input file.
- When writing, the `MAX_SIZE_README_LINE` is used to wrap text.
- All continuation lines are indented with 4 spaces for now, but that would be easy to change.
- The author list is split only between authors, I'm using chained `str.replace` to replace spaces with numbers to avoid wrapping within authors. It feels a bit like a hack, but it works for now.

I still need to add proper tests. Once that is done, I'll mark this as ready for review!

neutrinoceros commented:
Thanks a lot for working on this @vandalt !
I'm requesting a review from myself as I expect to be working rather intensely on MRT support in astropy in the coming months so it feels like I should give it a deep look at some point, if only to better familiarize myself.
However I'm not an expert yet on this region of the code base, so I don't think I can answer all your questions at the moment, but I do have one for you: do we already have a reason to go for `OrderedDict` ? I'm aware there are limitations with the builtin `dict` (which is why https://github.com/astropy/astropy/issues/10888 is still opened), but do we *know* that `dict` isn't sufficient her ?

vandalt commented:
Thanks for the feedback @neutrinoceros !

The only reason I went for `OrderedDict` is that [the docs for Table](https://docs.astropy.org/en/stable/table/construct_table.html#meta) mention that `meta` is an `OrderedDict` by default, so I went for that. Also, I had missed that `dict` is now officially ordered by default in Python 3.7+. I'll try switching to dict, everything should work.

vandalt committed:
Add proper test of metadata header roundtrip

vandalt committed:
Replace OrderedDict with dict in MRTHeader

vandalt committed:
[ci skip] Fix indentation of explanation continuation lines

neutrinoceros commented:
Oh, if that's how we document it, we should keep an `OrderedDict` here. Sorry about my misleading question !

vandalt committed:
Add explanations for extra space in explanation continuation lines

vandalt committed:
Remove TODO about ValueError vs RuntimeError

"An operation receives an argument with the right type but an
inappropriate value"
(https://docs.python.org/3/library/exceptions.html#ValueError)
seems to match the description.
KeyError would be if key not found, which is not the case here.

vandalt committed:
Remove TODOs about wrapping

vandalt committed:
Add tests for private metadata-reading methods of `MrtHeader`

Create new `test_mrt.py`, because this is not applicable to CDS tables

vandalt committed:
Handle corner cases in metadata-extraction and make tests pass

vandalt committed:
Move MRT-only tests to `test_mrt.py` instead of `test_cds.py`

vandalt commented:
I am encountering issues due to the various ways metadata other than byte-by-byte description can be specified in the header, namely article title, author names, and table titles

- The [CDS standard](http://vizier.u-strasbg.fr/doc/catstd-3.1.htx) has a given way to describe the metadata, as in [cdsFunctional.dat](https://github.com/astropy/astropy/blob/a9b25da67cc3239727a469a7116b75c01059ae7d/astropy/io/ascii/tests/data/cdsFunctional.dat?plain=1#L3)
- The MRT standard is not as explicit in its description, but the example at the bottom of [this page](https://journals.aas.org/mrt-standards/) and the paragraph just above seems to imply that we expect the `Title:`, `Authors:` and `Table:` fields with the format shown. This is what we have in example file [cds.dat](https://github.com/astropy/astropy/blob/a9b25da67cc3239727a469a7116b75c01059ae7d/astropy/io/ascii/tests/data/cds.dat?plain=1#L38), for example.

My current implementation supports only the latter case, as I was focusing on MRT tables. However, this means that now there are some tests failing when trying to read metadata from `cdsFunction.dat` with the MRT reader, for example.

What would be the best path forward:

1. Only support tables with the MRT-formatted metadata (as in cds.dat) with the MRT writer and update the tests and documentation to reflect this, then the CDS-formatted metadata can be handled in the CDS reader.
2. Support both the MRT-formatted and CDS-formatted metadata with the MRT reader, and update the MRT code I implemented to reflect this.
3. Other suggestions?

I feel like option 1 above separates the two formats nicely, but I might be missing some context about what is expected from the two standards. @taldcroft your suggestions would be welcome, since I think it was you who initially separated MRT and CDS in two classes.

Thanks!

hamogu commented:
It's things like this where @taldcroft and I always felt that the CDS and the MRT format are different. But then we recently talked to the AAS journal data editors and my understanding is that Gus Muench (who works as data editor for the AAS journal and makes the MRT tables they publish - @augustfly ) implied that they are not intended to be different. 
I'll point him by email to this conversation and ask for input; but if they really are different, then we need to have two different (and very similar) formats.

taldcroft commented:
@vandalt - seconding @neutrinoceros, thanks for working on this!

The question of maintaining newlines when reading in metadata is a good one. One complication is the lack of a real specification for the metadata. For instance, is it allowed to have two paragraphs with a blank line in between? Should newlines in general be considered significant?

For the Notes, my inclination would be to maintain the location of newlines. One reason would be an embedded equation that gets split badly if you rewrap to a new length. So maybe the simplest is to keep a string with the newlines, so a user can easily print it. The leading whitespace should be ignored.

taldcroft commented:
About CDS versus MRT, the separation of the two seems unambiguous to me. Although they have some commonalities in the data representation, most of the key metadata are represented in entirely different formats. Most notably the `Title`, `Table`, and `Authors` fields in  MRT have no analog in CDS. 

The MRT specification provides no guidance on whether fields like `ADC_Keywords` or `Description` are allowed, or whether arbitrary keywords `MyKeyword:` are allowed. Overall it would be beneficial to have a spec for MRT that is more complete.

### CDS
```
I/221               The Magellanic Catalogue of Stars - MACS (Tucholke+ 1996)
================================================================================
The Magellanic Catalogue of Stars - MACS
     Tucholke H.-J., de Boer K.S., Seitter W.C.
    <Astron. Astrophys. Suppl. Ser., 119, 91-98 (1996)>
    =1995Msngr..81...20D
================================================================================
ADC_Keywords: Magellanic Clouds ; Positional data

Description:
    The Magellanic Catalogue of Stars (MACS) is based on scans of ESO
    Schmidt plates and contains about 244,000 stars covering large areas
     ... 

File Summary:
--------------------------------------------------------------------------------
 FileName    Lrecl    Records    Explanations
--------------------------------------------------------------------------------
ReadMe          80          .    This file
lmc.dat         52     175779    The Large Magellanic Cloud
smc.dat         52      67782    The Small Magellanic Cloud
--------------------------------------------------------------------------------
Byte-by-byte Description of file: lmc.dat smc.dat
 ```
### MRT
```
Title: An Interesting Article submitted to the Research Notes of the AAS
Authors: Muench A.A., Schwarz G., Payne-Gaposchkin C., Banneker B., Hubble E.
Table: Radial Velocities of Unbound Interstellar Asteroids
================================================================================
Byte-by-byte Description of file: datafile1.txt
```

taldcroft commented:
About newlines in `Notes`, the example in https://journals.aas.org/mrt-explanation/ strongly suggests that newlines need to be maintained:
```
Note (2):
    C = CCD photometry (CfA redshift survey 1992)
    H = Zwicky magnitude split by observers
    K = Markarian catalogue of galaxies (1967)
    M = MCG (Vorontsov-Velyaminov et al. 1962-68)
    P = observer's eye estimate
    R = RC1 (de Vaucouleurs & de Vaucouleurs 1964)
```
In general, I would say that newlines should be maintained in all metadata fields with the exception of `Authors`, `Table`, and `Title`.

taldcroft commented:
Here are other questions I have about the MRT standard. I am using the word "field" for metadata values such as `Title` or `Keywords`. **These questions might best be answered in a formal specification document developed in collaboration with AAS**.

### Preamble section
```
Title: An Interesting Article submitted to the Research Notes of the AAS
Authors: Muench A.A., Schwarz G., Payne-Gaposchkin C., Banneker B., Hubble E.
Table: Radial Velocities of Unbound Interstellar Asteroids
```
- What fields are allowed in the preamble? Is there a list of recommended fields (e.g. `Description`, `Keywords`, `ADC_Keywords`)? Can authors provide any field names they choose?
- What are the requirements for continuation of multiline field values? Presumably any leading whitespace signals
continuation of the preceding line.
- Is a blank line within a field value allowed, e.g. for a `Description` with two paragraphs?
- For which fields should newlines in multiline values be preserved?

### Byte-by-byte Description section
This section is generally well-specified. However, I have not seen requirements for the separator lines. Is it required to have exactly the sequence of `====`, then three lines of `-----` as shown in the example? How long do those separator lines need to be?
```
================================================================================
Byte-by-byte Description of file: datafile1.txt
--------------------------------------------------------------------------------
   Bytes Format Units Label  Explanations
--------------------------------------------------------------------------------
   1-  8 F8.3   d       BJD      Barycentric Julian date, BJD-2454900, days
  10- 15 F6.2   m/s     RV       Radial Velocity (1)
  17- 20 F4.2   m/s   e_RV       1-sigma RV Uncertainty
  22- 26 F5.3   ---     SHK      S-value Activity Indicator
--------------------------------------------------------------------------------
```
### Notes section
- Are CDS-style notes like `Note on SHK:` allowed?
- Are other fields allowed in this section?
- What are the requirements for the end of Notes section separator line? All `----` again or is `====` allowed?
```
Note (1): Barycentric Radial velocity; see Lindegren & Dravins (2003A&A...401.1185L)
--------------------------------------------------------------------------------
```
### Epilogue section?
CDS allows a final "epilogue" section after the `Notes` like below that ends with `====`. Is this allowed in MRT?
```
Author’s address:
    Hans-Joachim Tucholke <tucholke@astro.uni-bonn.de>
================================================================================
```
### Data section
```
 456.789  -1.11 3.33 0.777
... data continues ...
```

vandalt committed:
Join notes with newline characters to preserve line splits

We don't want to split equations.
For mapped notes, the RHS is also joined with newline

vandalt committed:
Write notes and mapped notes assuming newlines are preserved

vandalt commented:
Thank you for the feedback!

> About CDS versus MRT, the separation of the two seems unambiguous to me. Although they have some commonalities in the data representation, most of the key metadata are represented in entirely different formats. Most notably the Title, Table, and Authors fields in MRT have no analog in CDS.

That is also my understanding. Currently a lot of the `read` tests use the CDS and MRT readers for the same files. This is fine for the data and byte-by-byte sections, but will not be for the metadata sections I am adding here. Also, writing is always done using `ascii.mrt`, from what I can tell. This will no longer work once the metadata is taken into account, as writing the metadata will be done differently. My current plan to get all tests passing is:

- Reorganize the `write_header()` code a bit so that the byte-by-byte section is shared, but other metadata is not.
- Reorganize the tests a bit so that CDS-formatted files are test with the CDS reader and MRT-formatted files are tested with the MRT reader. If there are tests that concern only the data or byte-by-byte part, I will keep both the MRT and CDS readers.

> The MRT specification provides no guidance on whether fields like ADC_Keywords or Description are allowed, or whether arbitrary keywords MyKeyword: are allowed. Overall it would be beneficial to have a spec for MRT that is more complete.

I agree. I am keeping the implementation simple with only `Title`, `Authors` and `Table` for now, but it should be straightforward to add other fields later on.

> CDS suggests header lines no longer than 70 characters. Is this the case for MRT?

Given that the goal is to facilitate conversion to FITS `COMMENT` cards, my best guess would be yes, but I could not find documentation on this. I guess it does not hurt to try and keep them under 70 on the astropy side?

taldcroft commented:
@augustfly has created a document based on my questions above, and this is informally evolving in the direction of a specification document. I added some content / suggestions:

https://docs.google.com/document/d/1rayUWoxbDbwIrNCdX3mpiOCg1kmjBlGgsIBgbbwuvSU/edit?usp=sharing

dhomeier commented:
> Oh, if that's how we document it, we should keep an `OrderedDict` here. Sorry about my misleading question !

No, changing this is correct, as we've overall been switching to `dict` where possible. The default for `Table` has in fact been changed in #16250 (i.e. for 7.0) – @taldcroft it seems the [dev docs](https://docs.astropy.org/en/latest/table/construct_table.html#meta) should perhaps be updated as well?

taldcroft commented:
> It seems the [dev docs](https://docs.astropy.org/en/latest/table/construct_table.html#meta) should perhaps be updated as well?

Yup, maybe:

The meta argument is an object that contains metadata associated with the table. It is recommended that this object be a dict ~~or OrderedDict~~, but the only firm requirement is that *it must be a dict-like mapping* and can be copied with the standard library copy.deepcopy() routine. By default, meta is an empty ~~OrderedDict~~dict.

dhomeier commented:
#17336

vandalt committed:
[ci skip] Add `write` methods to CDS and fix MRT writer for empty metadata

Byte-by-byte is exactly the same in CDS and MRT, but metadata is not

vandalt committed:
[ci skip] Move splitter and data-writing functionality from MRT to CDS

vandalt committed:
[ci-skip] Rename MRT-formatted files from `cds` to `mrt`

vandalt committed:
Test malformed CDS/MRT data with CDS as well as MRT

vandalt committed:
Add test for CDS in test_connect.py

vandalt committed:
[ci skip] Move tests common between MRT and CDS to a separate file with fixutre

vandalt committed:
Fix typo in TODO about max line width for CDS

vandalt committed:
[ci skip] remove "top metadata" lines from expected empty CDS output

vandalt committed:
[ci skip] Copy `cdsFunctional*.dat` files to mrtFunctional files to for tests

Now that metadata is not read in the same way, byte-by-byte tests need
separate files

neutrinoceros commented:
I realize I may have mis-communicated here, and maybe we are in a deadlock because of it. So, just in case you guys are waiting for my review, I should explicitly state that I don't normally review PRs marked as draft _unless explicitly requested_ (by someone else). In effect, I'm waiting for either the PR to be undrafted or that someone else asks me to review.
Sorry for this mishap !

vandalt commented:
Hi! Sorry, the issue is not miscommunication on your side, but lack of time on my end (I'm teaching this semester). I hope I can get back to this around May. However, if someone wants to take over the PR before that, please feel free to do so :)
